<a href="https://colab.research.google.com/github/Kaustubh-Atey/Keras-Codes-/blob/master/Intro_to_Keras_Basic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Creating a simple neural network using Keras for a binary classification task.**

 # **Data Import and Processing**


The data used here is : '**Pima Indians Diabetes Dataset**'. It is downloaded from : https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.csv

It is a binary (2-class) classification problem. There are 768 observations with 8 input variables and 1 output variable.

The variable names are as follows:

**1. Number of times pregnant.** 

**2. Plasma glucose concentration a 2 hours in an oral glucose tolerance test.** 

**3. Diastolic blood pressure (mm Hg).** 

**4. Triceps skinfold thickness (mm).** 

**5. 2-Hour serum insulin (mu U/ml).** 

**6. Body mass index (weight in kg/(height in m)^2).** 

**7. Diabetes pedigree function.** 

**8. Age (years).** 

**9. Class variable (0 or 1).** 


In [0]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import sklearn

In [45]:
# load data 
data = np.loadtxt('Pima Indians Diabetes Dataset.txt',delimiter =',')
np.shape(data)
data

array([[  6.   , 148.   ,  72.   , ...,   0.627,  50.   ,   1.   ],
       [  1.   ,  85.   ,  66.   , ...,   0.351,  31.   ,   0.   ],
       [  8.   , 183.   ,  64.   , ...,   0.672,  32.   ,   1.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,   0.245,  30.   ,   0.   ],
       [  1.   , 126.   ,  60.   , ...,   0.349,  47.   ,   1.   ],
       [  1.   ,  93.   ,  70.   , ...,   0.315,  23.   ,   0.   ]])

In [46]:
X_data = data[:,0:8]
print('X_data:',np.shape(X_data))
Y_data = data[:,8]
print('Y_data:',np.shape(Y_data))

X_data: (768, 8)
Y_data: (768,)


In [0]:
# Split data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X_data, Y_data, test_size=0.25, random_state= 0)

In [48]:
# Check the dimension of the sets
print('X_train:',np.shape(X_train))
print('y_train:',np.shape(y_train))
print('X_test:',np.shape(X_test))
print('y_test:',np.shape(y_test))

X_train: (576, 8)
y_train: (576,)
X_test: (192, 8)
y_test: (192,)


In [49]:
X_train

array([[  9.   ,  89.   ,  62.   , ...,  22.5  ,   0.142,  33.   ],
       [  1.   , 118.   ,  58.   , ...,  33.3  ,   0.261,  23.   ],
       [  0.   ,  91.   ,  80.   , ...,  32.4  ,   0.601,  27.   ],
       ...,
       [  4.   ,  94.   ,  65.   , ...,  24.7  ,   0.148,  21.   ],
       [ 11.   ,  85.   ,  74.   , ...,  30.1  ,   0.3  ,  35.   ],
       [  5.   , 136.   ,  82.   , ...,   0.   ,   0.64 ,  69.   ]])

# **Model**

In [0]:
import keras
from keras.models import Sequential   # importing Sequential model
from keras.layers import Dense        # importing Dense layers

In [0]:
# declaring model
basic_model = Sequential()

In [0]:
# Adding layers to the model
# First layers: 16 neurons/perceptrons that takes the input and uses 'sigmoid' activation function.
basic_model.add(Dense(units = 16 , activation = 'sigmoid', input_shape = (8,))) 
# Second layer: 1 neuron/perceptron that takes the input from the 1st layers and gives output as 0 or 1.Activation used is 'Hard Sigmoid'
basic_model.add(Dense(1, activation = 'hard_sigmoid'))

In [0]:
# compiling the model
sgd = keras.optimizers.SGD(lr=0.5, momentum=0.9, nesterov=True)
basic_model.compile(loss = 'binary_crossentropy', optimizer = 'sgd', metrics = ['accuracy'])

# **Training and Testing**

In [66]:
# training the model
basic_model.fit(X_train, y_train, epochs=100)

Epoch 1/100
576/576 [==============================] - 2s 3ms/step - loss: 0.6813 - acc: 0.6424
Epoch 2/100
576/576 [==============================] - 0s 60us/step - loss: 0.6740 - acc: 0.6424
Epoch 3/100
576/576 [==============================] - 0s 65us/step - loss: 0.6686 - acc: 0.6424
Epoch 4/100
576/576 [==============================] - 0s 56us/step - loss: 0.6645 - acc: 0.6424
Epoch 5/100
576/576 [==============================] - 0s 62us/step - loss: 0.6614 - acc: 0.6424
Epoch 6/100
576/576 [==============================] - 0s 66us/step - loss: 0.6590 - acc: 0.6424
Epoch 7/100
576/576 [==============================] - 0s 62us/step - loss: 0.6572 - acc: 0.6424
Epoch 8/100
576/576 [==============================] - 0s 60us/step - loss: 0.6559 - acc: 0.6424
Epoch 9/100
576/576 [==============================] - 0s 59us/step - loss: 0.6549 - acc: 0.6424
Epoch 10/100
576/576 [==============================] - 0s 59us/step - loss: 0.6542 - acc: 0.6424
Epoch 11/100
576/576 [========

In [67]:
# Test, Loss and accuracy
loss_and_metrics = basic_model.evaluate(X_test, y_test)
print('Loss = ',loss_and_metrics[0])
print('Accuracy = ',loss_and_metrics[1])

192/192 [==============================] - 1s 3ms/step
Loss =  0.6317572295665741
Accuracy =  0.6770833333333334


Accuracy of **0.677** was obtained from this very simple neural network. Using a deeper model and hyperparameter tuning will lead to better accuracy.

